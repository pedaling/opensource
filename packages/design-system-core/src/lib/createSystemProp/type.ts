import type { Theme } from '../types';

export type InterpolationFn = (props: Record<string, any>) => Record<string, any>;

export type SystemPropConfig = {
  property: string;
  styleProperty?: string;
  scale?: keyof Pick<Theme, 'colors' | 'space' | 'opacity'>;
  disabled?: boolean;
  transform?: (value: any) => Record<string, any> | undefined;
};

export type SystemProp = {
  (props: Record<string, any>): Record<string, any>;
  propName: string;
  disabled: boolean;
};
