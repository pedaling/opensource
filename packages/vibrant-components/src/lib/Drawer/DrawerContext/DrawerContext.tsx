import type { FC, ReactElement } from 'react';
import { createContext, useContext } from 'react';

type DrawerContextValue = {
  togglePanel: () => void;
  type: 'modal' | 'overlay' | 'standard';
  placement: 'bottom' | 'left' | 'right' | 'top';
  isOpen: boolean;
  deliverPanelSize: (panelSize: number) => void;
  containerSize: number;
};

const DrawerContext = createContext<DrawerContextValue>({
  togglePanel: () => {},
  type: 'standard',
  placement: 'right',
  isOpen: false,
  deliverPanelSize: () => {},
  containerSize: 0,
});

type DrawerProviderProps = {
  children: ReactElement | ReactElement[];
  togglePanel: () => void;
  type: 'modal' | 'overlay' | 'standard';
  placement: 'bottom' | 'left' | 'right' | 'top';
  isOpen: boolean;
  deliverPanelSize: (panelSize: number) => void;
  containerSize: number;
};

export const DrawerProvider: FC<DrawerProviderProps> = ({
  children,
  togglePanel,
  type,
  placement,
  isOpen,
  deliverPanelSize,
  containerSize,
}) => (
  <DrawerContext.Provider value={{ togglePanel, type, placement, isOpen, deliverPanelSize, containerSize }}>
    {children}
  </DrawerContext.Provider>
);

export const useDrawer = () => {
  const { togglePanel, type, placement, isOpen, deliverPanelSize, containerSize } = useContext(DrawerContext);

  return {
    togglePanel,
    type,
    placement,
    isOpen,
    deliverPanelSize,
    containerSize,
  };
};
