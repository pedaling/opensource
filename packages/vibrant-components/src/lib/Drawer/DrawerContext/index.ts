export { DrawerProvider, useDrawer } from './DrawerContext';
