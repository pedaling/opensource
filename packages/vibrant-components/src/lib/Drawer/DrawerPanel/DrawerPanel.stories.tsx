import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { DrawerPanel } from './DrawerPanel';

export default {
  title: 'DrawerPanel',
  component: DrawerPanel,
  args: {},
} as ComponentMeta<typeof DrawerPanel>;

export const Basic: ComponentStory<typeof DrawerPanel> = props => <DrawerPanel {...props} />;
