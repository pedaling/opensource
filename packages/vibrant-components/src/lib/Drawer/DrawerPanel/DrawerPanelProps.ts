import type { ReactElement } from 'react';
import type { ResponsiveValue } from '@vibrant-ui/core';
import { withVariation } from '@vibrant-ui/core';

export type PercentWidth = `${number}%`;

export type DefaultPanelSizeType = PercentWidth | number | 'auto';

export type DrawerPanelProps = {
  children: ReactElement[];
  testId?: string;
  defaultSize?: ResponsiveValue<DefaultPanelSizeType>;
};

export const withDrawerPanelVariation = withVariation<DrawerPanelProps>('DrawerPanel')();
