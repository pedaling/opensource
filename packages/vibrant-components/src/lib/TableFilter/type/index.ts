export type StringFilterOperator = 'contains' | 'empty' | 'equals' | 'notContains' | 'notEmpty' | 'notEquals';

export type StringFilter = {
  value: string;
  operator: StringFilterOperator[];
};

export type Filter = StringFilter & { dataKey: string };
