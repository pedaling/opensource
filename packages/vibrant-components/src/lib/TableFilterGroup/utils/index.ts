import type { DateFilterOperator, MultiSelectFilterOperator, Option, StringFilterOperator } from '../types';

export const isValueRequiredOperator = (op: string) => op === 'empty' || op === 'notEmpty';

export const isStringFilterValid = (filter: { value: string; operator: StringFilterOperator }) =>
  Boolean(filter.value) || isValueRequiredOperator(filter.operator);

export const isDateFilterValid = (filter: { value: Date[]; operator: DateFilterOperator }) => {
  if (isValueRequiredOperator(filter.operator)) return true;

  if (filter.operator === 'between') return filter.value.length >= 2;

  return filter.value.length >= 1;
};

export const isMultiSelectFilterValid = (filter: { value: Option['value'][]; operator: MultiSelectFilterOperator }) =>
  Boolean(filter.value.length > 0) || isValueRequiredOperator(filter.operator);
