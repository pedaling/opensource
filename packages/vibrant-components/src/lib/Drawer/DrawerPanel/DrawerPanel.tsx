import type { ReactElement } from 'react';
import { Children, useMemo } from 'react';
import type { LayoutEvent } from '@vibrant-ui/core';
import { Box } from '@vibrant-ui/core';
import { Transition } from '@vibrant-ui/motion';
import { Divider } from '../../Divider';
import { VStack } from '../../VStack';
import { useDrawer } from '../DrawerContext';
import { DrawerFooter } from '../DrawerFooter/DrawerFooter';
import { DrawerHeader } from '../DrawerHeader';
import type { DefaultPanelSizeType } from './DrawerPanelProps';
import { withDrawerPanelVariation } from './DrawerPanelProps';

const ANIMATE_DURATION = 200;

export const DrawerPanel = withDrawerPanelVariation(({ testId = 'drawer-panel', children, defaultSize = 360 }) => {
  const childArray = Children.toArray(children) as ReactElement[];

  const header = childArray.filter(child => child.type === DrawerHeader);
  const footer = childArray.filter(child => child.type === DrawerFooter);
  const contents = childArray.filter(child => child.type !== DrawerHeader && child.type !== DrawerFooter);

  const { placement, type, isOpen, containerSize, deliverPanelSize, panelSizePixel } = useDrawer();

  const isVertical = placement === 'left' || placement === 'right';

  const isStandardType = type === 'standard';

  const isStandardLeft = isStandardType && placement === 'left';

  const isStandardTop = isStandardType && placement === 'top';

  const onPanelLayout = ({ width, height }: LayoutEvent) => {
    const currentPanelSize = isVertical ? width : height;

    deliverPanelSize(currentPanelSize);
  };

  const trimPanelSize = (size: DefaultPanelSizeType, value: number) => {
    if (typeof size === 'string' && size !== 'auto') {
      const ratio = parseInt(size) * 0.01;

      return Number((ratio * value).toFixed(1));
    }

    return size;
  };

  const panelSize = useMemo(() => {
    if (Array.isArray(defaultSize)) {
      return (defaultSize as DefaultPanelSizeType[]).map(size => trimPanelSize(size, containerSize));
    }

    return trimPanelSize(defaultSize, containerSize);
  }, [containerSize, defaultSize]);

  const panelContent = isVertical ? (
    <Box width={panelSize} data-testid={testId} flexGrow={1} height="100%" onLayout={onPanelLayout} flexDirection="row">
      {!isStandardLeft && <Divider direction="vertical" kind="default" />}
      <VStack height="100%" alignVertical="space-between">
        <Box width="100%" overflow="hidden">
          {header}
          <Box>{contents}</Box>
        </Box>
        <Box width="100%">{footer}</Box>
      </VStack>
      {isStandardLeft && <Divider direction="vertical" kind="default" />}
    </Box>
  ) : (
    <Box height={panelSize} width="100%" data-testid={testId} onLayout={onPanelLayout} flexDirection="column">
      {!isStandardTop && <Divider direction="horizontal" kind="default" />}
      <VStack height="100%" alignVertical="space-between">
        <Box width="100%" overflow="hidden">
          {header}
          <Box>{contents}</Box>
        </Box>
        <Box width="100%">{footer}</Box>
      </VStack>
      {isStandardTop && <Divider direction="horizontal" kind="default" />}
    </Box>
  );

  return isStandardType ? (
    <>
      {placement === 'right' && (
        <Box position="absolute" top={0} right={0} height="100%" flexDirection="row">
          <Transition
            animation={{
              x: isOpen ? -panelSizePixel : 0,
            }}
            duration={ANIMATE_DURATION}
            easing="easeOutQuad"
          >
            <Box width={panelSize} right={-panelSizePixel} height="100%">
              {panelContent}
            </Box>
          </Transition>
        </Box>
      )}
      {placement === 'left' && (
        <Box position="absolute" top={0} left={0} height="100%" flexDirection="row">
          <Transition
            animation={{
              x: isOpen ? panelSizePixel : 0,
            }}
            duration={ANIMATE_DURATION}
            easing="easeOutQuad"
          >
            <Box width={panelSize} left={-panelSizePixel} height="100%">
              {panelContent}
            </Box>
          </Transition>
        </Box>
      )}
      {placement === 'top' && (
        <Transition
          animation={{
            y: isOpen ? panelSizePixel : 0,
          }}
          duration={ANIMATE_DURATION}
          easing="easeOutQuad"
        >
          <Box height={panelSize} top={-panelSizePixel}>
            {panelContent}
          </Box>
        </Transition>
      )}
      {placement === 'bottom' && (
        <Transition
          animation={{
            height: isOpen ? panelSize : 0,
          }}
          duration={ANIMATE_DURATION}
          easing="easeOutQuad"
        >
          <Box height={panelSize}>{panelContent}</Box>
        </Transition>
      )}
    </>
  ) : (
    <>
      {placement === 'right' && (
        <Transition animation={{ x: isOpen ? -panelSizePixel : 0 }} duration={ANIMATE_DURATION} easing="easeOutQuad">
          <Box
            height="100%"
            backgroundColor="surface2"
            position="absolute"
            top={0}
            right={-panelSizePixel}
            zIndex={3}
            elevationLevel={isOpen ? 3 : undefined}
          >
            {panelContent}
          </Box>
        </Transition>
      )}
      {placement === 'left' && (
        <Transition animation={{ x: isOpen ? panelSizePixel : 0 }} duration={ANIMATE_DURATION} easing="easeOutQuad">
          <Box
            height="100%"
            backgroundColor="surface2"
            position="absolute"
            top={0}
            left={-panelSizePixel}
            zIndex={3}
            elevationLevel={isOpen ? 3 : undefined}
          >
            {panelContent}
          </Box>
        </Transition>
      )}
      {placement === 'top' && (
        <Transition animation={{ y: isOpen ? panelSizePixel : 0 }} duration={ANIMATE_DURATION} easing="easeOutQuad">
          <Box
            width="100%"
            backgroundColor="surface2"
            position="absolute"
            top={-panelSizePixel}
            left={0}
            zIndex={3}
            elevationLevel={isOpen ? 3 : undefined}
          >
            {panelContent}
          </Box>
        </Transition>
      )}
      {placement === 'bottom' && (
        <Transition animation={{ y: isOpen ? -panelSizePixel : 0 }} duration={ANIMATE_DURATION} easing="easeOutQuad">
          <Box
            width="100%"
            backgroundColor="surface2"
            position="absolute"
            bottom={-panelSizePixel}
            left={0}
            zIndex={3}
            elevationLevel={isOpen ? 3 : undefined}
          >
            {panelContent}
          </Box>
        </Transition>
      )}
    </>
  );
});
