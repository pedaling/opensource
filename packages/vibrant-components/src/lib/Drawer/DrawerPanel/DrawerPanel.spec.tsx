import { Box, Text } from '@vibrant-ui/core';
import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { DrawerProvider } from '../DrawerContext';
import { DrawerPanel } from './DrawerPanel';

describe('<DrawerPanel />', () => {
  const { render } = createReactRenderer();

  let renderer: ReactRenderer;

  describe('when component is rendering', () => {
    beforeEach(async () => {
      renderer = render(
        <DrawerPanel defaultSize="30%">
          <Box />
        </DrawerPanel>
      );
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('When size is empty', () => {
    let element: HTMLElement;

    beforeEach(async () => {
      renderer = render(
        <DrawerProvider
          placement="right"
          type="overlay"
          isOpen={true}
          containerSize={1200}
          deliverPanelSize={() => {}}
          panelSizePixel={360}
          togglePanel={() => {}}
        >
          <DrawerPanel>
            <Text>Default is 360</Text>
          </DrawerPanel>
        </DrawerProvider>
      );

      element = await renderer.findByTestId('drawer-panel');
    });

    it('default size is 360', () => {
      expect(element).toHaveStyleRule('width', '360px');
    });
  });
});
