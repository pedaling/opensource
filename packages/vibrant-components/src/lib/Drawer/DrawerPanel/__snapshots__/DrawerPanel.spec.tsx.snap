// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawerPanel /> when component is rendering match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  position: absolute;
  top: 0;
  right: 0;
  height: 100%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 0;
  right: 0;
  height: 100%;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  position: relative;
  width: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 100%;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  border-width: 0;
  border-color: #00000026;
  border-style: solid;
  width: 0;
  border-right-width: 1px;
  height: 100%;
  margin-left: 0;
  margin-right: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  height: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 100%;
  overflow: hidden;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 100%;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
      style="transform: translateX(0px); transition-property: transform; transition-duration: 200ms; transition-timing-function: cubic-bezier(0.5, 1, 0.89, 1);"
    >
      <div
        class="emotion-2"
        data-testid="drawer-panel"
      >
        <div
          class="emotion-3"
          kind="default"
        />
        <div
          class="emotion-4"
        >
          <div
            class="emotion-5"
          >
            <div
              class="emotion-6"
            >
              <div
                class="emotion-6"
              />
            </div>
          </div>
          <div
            class="emotion-8"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
